src/charm.py

```diff
@@ -92,9 +92,11 @@ def mysql(self) -> MySQL:
         return MySQL(mysql_config)
 
     @property
-    def unit_ip(self) -> str:
+    def unit_ip(self) -> Optional[str]:
```

Comment:
Why not always return a string. If the bind address is not available then return an empty string that is false in a boolean context.

Reply:
I think a `None` is safer: no need to worry about carefully manipulating the returned value - if mishandled it will raise an exception, rather than doing something unpredictable, for example appending a port number to an empty string.

Reply:
Weather `None` or empty string, you still need to check if you got a valid value either by a predicate or by trapping and exception.

Reply:
I meant that `None` is more fail safe:
If you're checking output then all is good; if one may forgot to check output, then None is preferable.
Also, the `Optional` annotation conveys very clearly to the user that the call may "fail".

Reply:
Yes, it's better

Reply:
This is a good example of code that should be in the framework...